File: interpreter/interpreter/interpreter.py
from .parser import Parser
from .ast import Number, BinOp, UnarOp

class NodeVisitor:
    
    def visit(self):
        pass

class Interpreter(NodeVisitor):
    
    def __init__(self):
        self.parser = Parser()

    def visit(self, node):
        if isinstance(node, Number):
            return self.visit_number(node)
        elif isinstance(node, BinOp):
            return self.visit_binop(node)
        elif isinstance(node, UnarOp):
            return self.visit_unarop(node)

    def visit_number(self, node):
        return float(node.token.value)

    def visit_binop(self, node):
        match node.op.value:
            case "+":
                return self.visit(node.left) + self.visit(node.right)
            case "-":
                return self.visit(node.left) - self.visit(node.right)
            case "*":
                return self.visit(node.left) * self.visit(node.right)
            case "/":
                return self.visit(node.left) / self.visit(node.right)
            case _:
                raise ValueError("Invalid operator")

    def visit_unarop(self, node):
        match node.op.value:
            case "+":
                return self.visit(node.node)
            case "-":
                return -self.visit(node.node)
            case _:
                raise ValueError("Invalid operator")

    def eval(self, code):
        tree = self.parser.parse(code)
        return self.visit(tree)

File: pascalInterpreter/src/main/kotlin/interpreter/Interpreter.kt
package interpreter

class Interpreter {
    private val parser = Parser()
    val variables = mutableMapOf<String, Double>()

    fun visit(node: Node): Double {
        when (node) {
            is Number -> return visitNumber(node)
            is BinOp -> return visitBinOp(node)
            is Variable -> visitVariable(node)
            is Semicolon -> visitSemicolon(node)
            is Empty -> {}
            is Identifier -> return visitIdentifier(node)
            is UnarOp -> return visitUnarOp(node)
        }
        return 0.0
    }

    fun visitNumber(number: Number): Double {
        return number.token.value.toDouble()
    }

    fun visitBinOp(node: BinOp): Double {
        return when (node.op.value) {
            "+" -> visit(node.left) + visit(node.right)
            "-" -> visit(node.left) - visit(node.right)
            "*" -> visit(node.left) * visit(node.right)
            "/" -> visit(node.left) / visit(node.right)
            else -> throw IllegalArgumentException("Invalid operator")
        }
    }


    fun visitUnarOp(node: UnarOp): Double {
        return when (node.op.value) {
            "+" -> +visit(node.value)
            "-" -> -visit(node.value)
            else -> throw IllegalArgumentException("Invalid operator")
        }
    }

    fun visitIdentifier(node: Identifier): Double {
        if (node.token.value in variables.keys) {
            return variables[node.token.value]!!
        }
        throw IllegalArgumentException("Unknown variable")
    }

    fun visitSemicolon(node: Semicolon) {
        visit(node.left)
        visit(node.right)
    }

    fun visitVariable(node: Variable) {
        if (node.token.value !in variables.keys) {
            variables[node.token.value] = 0.0
        }
        variables[node.token.value] = visit(node.value)
    }

    fun eval(code: String): Double {
        val tree = parser.parse(code)
        return visit(tree)
    }
}